### DISCUSSION TITLE
How to Serialize DateTime as DateTimeKind.Unspecified?

### DISCUSSION THREAD

**Discussion by Molinware** [CoreWCF/CoreWCF#1344] [Q&A]
```
[DataContract]
public class TestClass
{
    [DataMember]
    public DateTime CurrentTime { get; set; }
}
```

In the case that the Server is GMT-2 and the Client is GMT-3.
When I send the date to the server it adds one hour.

Wanted behavior:
I want to send from the client 28/02/2024 15:43 and when gets to the server remains 28/02/2024 15:43.

Actual behavior:
Sending from the client 28/02/2024 15:43 and when gets to the server it changes to 28/02/2024 16:43.

Serializing with DateTimeKind.Unspecified would resolve the issue. I want to know if I can make all my Operations Serialize as DateTimeKind.Unspecified, or if there's another alternative to ignore GMT.

**Comment by mconnew** [comment]
I believe Unspecified is what you would need. You can experiment and try things out as the underlying implementation uses [XmlConvert.ToDateTime](https://learn.microsoft.com/en-us/dotnet/api/system.xml.xmlconvert.todatetime?view=net-8.0#system-xml-xmlconvert-todatetime(system-string-system-xml-xmldatetimeserializationmode)) passing XmlDateTimeSerializationMode.RoundtripKind as the serialization mode. To convert to a string, it uses [XmlConvert.ToString](https://learn.microsoft.com/en-us/dotnet/api/system.xml.xmlconvert.tostring?view=net-8.0#system-xml-xmlconvert-tostring(system-datetime-system-xml-xmldatetimeserializationmode)) passing the DateTime object along with the same RoundtripKind enum value. Hopefully this should be enough to experiment with for you to get confidence it will do what you want it to do.

**Reply by Molinware** [reply]
Is there a way to Serialize all DateTime's as DateTimeKind.Unspecified for all services?

**Comment by mconnew** [comment]
I remembered that things are done differently when using binary XML, which happens by default with any bindings with the name starting with Net, e.g. NetTcpBinding, NetHttpBinding. In those cases, it uses DateTime.ToBinary() and DateTime.FromBinary() which for Unspecified just keeps the bits identical in memory as it's encoded to a 64 bit number so will be 100% identical between the two ends. If you use Local, it converts it to UTC first and changes the time by the local offset before serializing as a 64 bit integer, so that would get the value changed. But Unspecified is safe.  

I quickly tested things using [CSharpRepl](https://github.com/waf/CSharpRepl) to verify the text case:
```shell
> var dt = new DateTime(2024, 03, 05, 0, 0, 0, DateTimeKind.Unspecified);

> XmlConvert.ToString(dt, XmlDateTimeSerializationMode.RoundtripKind)
"2024-03-05T00:00:00"
> XmlConvert.ToDateTime("2024-03-05T00:00:00", XmlDateTimeSerializationMode.RoundtripKind)
3/5/2024 12:00:00 AM
> XmlConvert.ToDateTime("2024-03-05T00:00:00", XmlDateTimeSerializationMode.RoundtripKind).Kind
Unspecified
> dt = new DateTime(2024, 03, 05, 0, 0, 0, DateTimeKind.Local);

> XmlConvert.ToString(dt, XmlDateTimeSerializationMode.RoundtripKind)
"2024-03-05T00:00:00-08:00"
## Changed the timezone offset to -07:00 to simulate server and client being in different timezones
> XmlConvert.ToDateTime("2024-03-05T00:00:00-07:00", XmlDateTimeSerializationMode.RoundtripKind)
3/4/2024 11:00:00 PM
> XmlConvert.ToDateTime("2024-03-05T00:00:00-07:00", XmlDateTimeSerializationMode.RoundtripKind).Kind
Local
```

That looks like using Unspecified does exactly what you want

**Reply by mconnew** [reply]
I reread your request and I think there's an aspect I might have missed, but I'm not sure. If you don't control the creation of the `DateTime` objects, you could convert them to Unspecified by using an `ISerializationSurrogateProvider`. I provided a basic example of using one [with the WCF Client here](https://github.com/dotnet/wcf/issues/5437#issuecomment-1970019073). It's similar for CoreWCF, but you would use the extension method `ConfigureServiceHostBase<TService>` and use `ServiceHostBase.Description` to get to the endpoints to modify the `DataContractSerializerOperationBehavior`. In the surrogate provider, if the instance is a `DateTime`, you would create a new one based off the provided one and make it `Unspecified`.

**Reply by Molinware** [reply]
Even if the return of the operation is a complex object the "GetObjectToSerialize" would go throught all nodes?
Or is the parameter "object obj" the complex object and I need to loop throught it?

**Reply by mconnew** [reply]
Good question, and I'm actually not sure. The complex object will definitely get passed to it, but so would each property. So if you had this:
```
class Foo
{
    public Bar MyBar { get; set; }
}
```
It would pass you Foo, which you would have the chance to replace with a new instance (it passes the real instance so if you make changes, you are changing the real Foo instance so generally better not to modify but create a new instance). If you just return the same Foo back, it will then pass in Bar and you can do the same thing. I'm not sure about `DateTime` though as it's considered a "primitive" and primitives have special handling in lots of places in the serializer. My best guess is that I think it would still pass you a DateTime though as the primitive aspect is that it's an XML primitive so a custom DataContract based on reflecting over it isn't needed, and no custom XML schema needs to be exported. You might still want to replace a DateTime with something else, which is why I believe it probably will. But I would only be 90% surprised if it doesn't. Try it to verify. Worst case is you need to reflect over all properties and modify any DateTime properties, but you won't have to do it recursively if you have to do it at all.

DateTime is a value type (a struct in c# nomenclature), so it gets boxed onto the heap as the method takes an object. Value types are supposed to be immutable, but just to cover my bases, if you are dealing with a mutable struct, modifying it in this method won't modify the original as structs are copied when passed, and copied when boxed/unboxed.
